src/fsharp/FSharp.Core/array.fs

```diff
@@ -483,19 +483,45 @@ namespace Microsoft.FSharp.Collections
                             count <- count + 1
             Array.subUnchecked 0 count res
 
-
         [<CompiledName("Filter")>]
-        let filter f (array: _[]) = 
-            checkNonNull "array" array
-            let res = Array.zeroCreateUnchecked array.Length 
-            let mutable count = 0
-            for x in array do                 
-                if f x then 
-                    res.[count] <- x
-                    count <- count + 1
-            Array.subUnchecked 0 count res
-
+        let filter f (array: _[]) =         
+            checkNonNull "array" array                    
+            let mutable i = 0    
+            while i < array.Length && not (f array.[i]) do
+                i <- i + 1
+            
+            if i <> array.Length then                    
+                let mutable element = array.[i]
+                let chunk1 : 'T[] = Array.zeroCreateUnchecked (((array.Length-i) >>> 2) + 1)
```

Comment:
Sometimes a little faster, sometimes a little slower, depending on how much gets filtered.
Big memory penalty when many things get filtered, which also causes a speed penalty. It does restore the fast path for when nothing gets filtered.

```ini

Host Process Environment Information:
BenchmarkDotNet=v0.9.8.0
OS=Microsoft Windows NT 6.2.9200.0
Processor=Intel(R) Core(TM) i7-4712HQ CPU 2.30GHz, ProcessorCount=8
Frequency=2240910 ticks, Resolution=446.2473 ns, Timer=TSC
CLR=MS.NET 4.0.30319.42000, Arch=64-bit RELEASE [RyuJIT]
GC=Concurrent Workstation
JitModules=clrjit-v4.6.1590.0

Type=CoreBenchmark  Mode=Throughput  Platform=X64  
Jit=RyuJit  GarbageCollection=Concurrent Workstation  

```
  Method | Length | Pct |      Median |     StdDev | Scaled | Gen 0 | Gen 1 |    Gen 2 | Bytes Allocated/Op |
-------- |------- |---- |------------ |----------- |------- |------ |------ |--------- |------------------- |
     **New** | **100000** |   **0** | **192.6800 us** |  **3.7269 us** |   **0.99** |     **-** |     **-** |        **-** |             **223.45** |
 OldPlus | 100000 |   0 | 194.2310 us |  5.4061 us |   1.00 |     - |     - |        - |             222.20 |
     **New** | **100000** |   **2** | **286.0879 us** |  **5.4825 us** |   **0.86** |     **-** |     **-** |   **254.91** |          **43,240.30** |
 OldPlus | 100000 |   2 | 330.7581 us |  5.6577 us |   1.00 |     - |     - | 1,098.28 |         169,005.43 |
     **New** | **100000** |  **10** | **371.8050 us** |  **3.3757 us** |   **0.92** |     **-** |     **-** |   **278.70** |          **64,468.67** |
 OldPlus | 100000 |  10 | 404.1597 us |  5.3421 us |   1.00 |     - |     - | 1,101.76 |         188,457.16 |
     **New** | **100000** |  **25** | **541.6640 us** | **10.5468 us** |   **0.92** |     **-** |     **-** |   **561.51** |          **88,952.07** |
 OldPlus | 100000 |  25 | 589.0407 us | 58.1685 us |   1.00 |     - |     - | 1,182.97 |         220,629.98 |
     **New** | **100000** |  **50** | **790.0072 us** | **13.8439 us** |   **1.00** |     **-** |     **-** | **1,322.00** |         **204,549.69** |
 OldPlus | 100000 |  50 | 787.2796 us |  6.5840 us |   1.00 |     - |     - | 1,576.00 |         244,716.71 |
     **New** | **100000** |  **75** | **656.1230 us** | **13.2718 us** |   **1.19** |     **-** |     **-** | **1,624.51** |         **254,655.45** |
 OldPlus | 100000 |  75 | 549.9804 us |  6.2456 us |   1.00 |     - |     - | 1,532.54 |         252,263.44 |
     **New** | **100000** |  **90** | **496.8645 us** | **10.6528 us** |   **1.01** |     **-** |     **-** | **1,956.62** |         **311,019.45** |
 OldPlus | 100000 |  90 | 490.2044 us | 40.9234 us |   1.00 |     - |     - | 1,775.84 |         315,199.39 |
     **New** | **100000** |  **98** | **481.1661 us** | **10.2857 us** |   **1.05** |     **-** |     **-** | **2,196.75** |         **327,805.23** |
 OldPlus | 100000 |  98 | 456.8940 us |  9.4714 us |   1.00 |     - |     - | 2,296.62 |         343,529.09 |
     **New** | **100000** | **100** | **474.2035 us** | **15.7986 us** |   **1.58** |     **-** |     **-** | **2,364.10** |         **357,264.53** |
 OldPlus | 100000 | 100 | 299.9004 us |  6.8316 us |   1.00 |     - |     - | 1,112.89 |         168,271.76 |